scikeras/wrappers.py

```diff
@@ -674,22 +670,25 @@ def _fit(self, X, y, sample_weight, warm_start, epochs, initial_epoch):
         else:
             X, y = self._validate_data(X, y)
 
-        X, y, sample_weight = self._validate_sample_weight(X, y, sample_weight)
+        if sample_weight is not None:
+            X, y, sample_weight = self._validate_sample_weight(X, y, sample_weight)
 
         y = self.target_encoder_.transform(y)
         X = self.feature_encoder_.transform(X)
 
         self._check_model_compatibility(y)
 
-        # fit model
-        return self._fit_keras_model(
-            X,
-            y,
-            sample_weight=sample_weight,
-            warm_start=warm_start,
-            epochs=epochs,
-            initial_epoch=initial_epoch,
-        )
+        # fit model if a target was given
+        if y is not None:
+            self._fit_keras_model(
```

Comment:
Why is model fitting performed only if a target is given?

Reply:
As far as I know, the only case in which you'd want to fit the Keras model without a target is when the `X` is a `tf.DataSet` (in which case the target is contained in `X`).

Can you think of any other situation where it would be valid to fit the actual model with `y=None`?

Reply:
And for `tf.DataSet`s, if we wanted to support them in the future, it would just be `if y is not None or isinstance(X, ...)`...

Reply:
Yes. Any unsupervised learning example. Examples include Scikit-learn's [TruncatedSVD][1], [NMF][2], and [PCA]. I'm sure Keras examples of those models can be created.

> As far as I know, the only case in which you'd want to fit the Keras model without a target is when the `X` is a `tf.DataSet`

So... `self._fit_keras_model` should be called if `y=None` (and `X` is a TF dataset)?

[1]:https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.TruncatedSVD.html#sklearn.decomposition.TruncatedSVD.fit
[2]:https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.NMF.html#sklearn.decomposition.NMF.fit
[3]:https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn.decomposition.PCA.fit

Reply:
> So... `self._fit_keras_model` should be called if `y=None` (and `X` is a TF dataset)?

In that case, yes, that's how you call a Keras model with a `tf.DataSet`. But that's definitely for another PR.


> Yes. Any unsupervised learning example. Examples include Scikit-learn's [TruncatedSVD](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.TruncatedSVD.html#sklearn.decomposition.TruncatedSVD.fit), [NMF](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.NMF.html#sklearn.decomposition.NMF.fit), and [PCA]. 

Wouldn't those all be implemented similar to how I did here? That is, `BaseWrapper.fit` _is_ called with `fit(X=..., y=None)` but `tf.keras.Model` is never called with `y=None` since it is trained separately.

Reply:
I do think there is the possibility of cleaning up the need to have two models with shared layers if SciKeras could somehow handle multiple models and allow using one for the initial `fit` and another for `predict/transform`, but I don't know that there's a universal way to support that use case. If there is, maybe we can make a `KerasTransformer` class to make the implementation in this notebook/PR cleaner.

Reply:
After implementing this specific example, I imagine the following:

```python3
def get_fit_model() -> tf.keras.Model:
    """A regular SciKeras get_model"""
    ...

def get_tf_model(fit_model: tf.keras.Model) -> tf.keras.Model:
    """A regular SciKeras get_model"""
    # extract layers from fitted model?
    ...

def get_inverse_tf_model(fit_model: tf.keras.Model) -> tf.keras.Model:
    """A regular SciKeras get_model"""
    # extract layers from fitted model?
    ...


class KerasTransformer(BaseEstimator, TransformerMixin):

    def __init__(self, model, tf_model, inverse_tf_model):
       ...

    def fit(self, X, y):
        super().fit(X, y)
        self.tf_model_ = get_tf_model(self.model_)
        self.inverse_tf_model_ = get_tf_model(self.model_)
        return self
    
    def transform(self, X):
        # some preprocessing (same as predict, etc.)
        X_tf = self.tf_model_.predict(X)
        # some postprocessing? (new methods/transformers?)
        return X_tf
    
    def inverse_transform(self, X_tf):
        # some preprocessing? (new methods/transformers?)
        X = self.inverse_tf_model_.predict(X_tf)
        # some postprocessing (same as predict, etc.?)
        return X
```

Reply:
Yes:
```
ValueError: No gradients provided for any variable: ...
```

Reply:
> ValueError: No gradients provided for any variable: ...

Right. Then maybe `y` should be required; the model always has an output (even it's an approximation of the input).

If not,  I think this code would satisfy both our concerns:

``` python
self._fit_keras_model(X=X, y=y or X, ...)
```

* `self._fit_keras_model` would always be called with `fit` (which I think is critical, and the reason for `fit`).
* It would work with autoencoders/unsupervised models where the input and output should match (an appropriate documentation note needs to added).
* It would allow setting `y=None`. If `X` is a dataset, it could be modified to the following:

``` python
self._fit_keras_model(X=X, y=(y or X) if not isinstance(X, tf.Dataset) else y, ...)
```

Reply:
What is the advantage to this proposed method over the current method in this PR? The `Dataset` part can be handled exactly the same way in both.

This also won't work with the current PR. We need to call `BaseWrapper.fit` and _not_ call `BaseWrapper.model_.fit` in order for initialization to happen before you can call `predict/transform`. I considered creating a separate method `initialize` that could be called instead of `BaseWrapper.fit`, but that would change the apparent API vs. sklearn transformers.

Reply:
> What is the advantage to this proposed method over the current method in this PR?

`self._fit_keras_model` is called when `y=None`. Why would you call `BaseWrapper.fit` and not fit a Keras model?

Reply:
To initialize internal stuff so that it is available for `predict`/`transform`, including:
1) Get/create Keras model. In this case it's a pretrained model used only for inference, not need to train it.
2) Instantiate feature transformers.
3) Learn input shapes and other data validation stuff.

Reply:
Which of the options below we trying to support?

* Allow for unsupervised models (e.g., `autoencoder.fit(X_train)`).
* Allow for pretrained models – `autoencoder.initialize(model=pretrained).predict(X_test)`.

If Keras requires that features `X` and targets `y` be passed to any model, I think forcing `y` to be specified is reasonable.

Reply:
Hmm. I think we should support both use cases.

For autoencoder based transformers specifically, unless we create a separate class like described in https://github.com/adriangb/scikeras/pull/123#discussion_r518244187, we need to have multiple models (like implemented in the tests/notebooks in this PR). For the non-training models (the models that are only used for inference), we need to be able to "initialize" everything by calling `fit` so that the sklearn interface works (`fit_transform`, etc.). Although it is an internal implementation, I do see how it could seem arbitrary to not call `model_.fit` if `y=None`.

Maybe we should:
1. Implement a `BaseWrapper.initialize` method that will pull together `BaseWrapper._initialize` and `BaseWrapper._validate_data` and make them public, with the purpose of supporting the general use case of pre-trained models. This would be a separate PR.
2. In these examples, implement an `AutoEncoderTransformer` class that looks something like https://github.com/adriangb/scikeras/pull/123#discussion_r518244187 but internally uses the `initialize` method. This would allow the external API to still fit into the sklearn API (`autoencoder.fit_transform(X)`) while avoiding the need for the `if y is not None` logic inside `BaseWrapper._fit`. I would not implement it as part of scikeras itself yet, since I am not convinced that a clean, general implementation is possible.

Does this sound good to you?

Reply:
@stsievert I implemented point (1) above. There are basically no changes to `_fit` now (aside from reshuffling logic into the public `initialize` method). This includes no `if y is not None` logic. While point (2) would be nice, I don't think it is required for this PR.